### PR TITLE
Add v4 support for images and figures

### LINKED ISSUE #3049: v4: Images
**Read first!** [Contributing](https://github.com/react-bootstrap/react-bootstrap/blob/master/CONTRIBUTING.md) | [Structuring a PR](https://github.com/react-bootstrap/react-bootstrap/blob/v4-support/V4-Contributing.md ) | [Bootstrap docs](https://getbootstrap.com/docs/4.0/content/images/)

### PR DESCRIPTION
**Pull request by iamnapo**
Fixes #3049, #3048

~~Opened as [WIP] to get feedback, will squash commits when done.~~

### CODE REVIEW

**Comment by iamnapo** [comment]
@jquense I'm not sure how you want me to handle `.figure-img`.

**Comment by iamnapo** [comment]
@jquense I'm not sure how you want me to handle `.figure-img`.

File: src/Image.js [hunk 1]
```diff
@@ -27,28 +27,28 @@ const propTypes = {
...
     } = this.props;
     const [bsProps, elementProps] = splitBsProps(props);
 
     const classes = {
-      [prefix(bsProps, 'responsive')]: responsive,
-      [prefix(bsProps, 'rounded')]: rounded,
-      [prefix(bsProps, 'circle')]: circle,
+      [prefix(bsProps, 'fluid')]: fluid,
+      rounded,
+      'rounded-circle': roundedCircle,
```
Comment (jquense) [inline src/Image.js]: I wonder if i's better to remove the `rounded` prop completely and point people to https://getbootstrap.com/docs/4.0/utilities/borders/ in the docs, which is what upstream is doing.

what do you think @taion ?

**Comment by jquense** [comment]
For `figure` i would make a new `Figure` component with Caption and Image sub components. Definitely write the FigureImage in terms of this Image component tho :)

File: src/FigureImage.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    const classes = {
+      'figure-img': figureImg,
+      [prefix(bsProps, 'fluid')]: fluid,
+      rounded,
+      'rounded-circle': roundedCircle,
+      [prefix(bsProps, 'thumbnail')]: thumbnail
+    };
+
+    return (
+      <img // eslint-disable-line jsx-a11y/alt-text
```
Comment (jquense) [inline src/FigureImage.js]: This is almost exactly the same as Image, I would write this one in terms of the other. So have FigureImage render a n Image but add the `figure-img` class to it

File: src/Image.js [hunk 3]
```diff
@@ -6,49 +6,49 @@ import { bsClass, prefix, splitBsProps } from './utils/bootstrapUtils';
...
-      responsive,
+      fluid,
       rounded,
-      circle,
+      roundedCircle,
       thumbnail,
       className,
       ...props
     } = this.props;
     const [bsProps, elementProps] = splitBsProps(props);
```
Comment (jquense) [inline src/Image.js]: Lets use the newer `createBootstrapComponent` recently added, and no bootstrapUtils. You can check out Button.js for a reference on how to use it!

**Comment by iamnapo** [comment]
@jquense thoughts?

File: src/FigureImage.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+import { createBootstrapComponent } from './ThemeProvider';
+
+import Image from './Image';
+
+class FigureImage extends React.Component {
+  static propTypes = { ...Image.propTypes };
+
+  static defaultProps = { ...Image.propTypes, fluid: true };
```
[the remark this answers is not in the document]
Comment (iamnapo) [reply src/FigureImage.js]: how can I override `fluid` then?

[the remark this answers is not in the document]
Comment (jquense) [reply src/FigureImage.js]: `static defaultProps = { fluid: true };` is enough :)

**Comment by iamnapo** [comment]
Thanks for the help!

**Comment by jquense** [comment]
Thanks a ton!

**Pull request merged**